lib/srv/usermgmt_unix.go

```diff
@@ -0,0 +1,95 @@
+//go:build !windows
+// +build !windows
+
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package srv
+
+import (
+	"os/user"
+
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+)
+
+/*
+#include <sys/types.h>
+#include <pwd.h>
+*/
+import "C"
+
+var _ HostUsersBackend = &UnixHostUsersBackend{}
+
+type UnixHostUsersBackend struct{}
+
+// Lookup implements host user information lookup
+func (*UnixHostUsersBackend) Lookup(username string) (*user.User, error) {
+	return user.Lookup(username)
+}
+
+// UserGIDs returns the list of group IDs for a user
+func (*UnixHostUsersBackend) UserGIDs(u *user.User) ([]string, error) {
+	return u.GroupIds()
+}
+
+// LookupGroup host group information lookup
+func (*UnixHostUsersBackend) LookupGroup(name string) (*user.Group, error) {
+	return user.LookupGroup(name)
+}
+
+// GetAllUsers returns a full list of users present on a system
+func (*UnixHostUsersBackend) GetAllUsers() ([]string, error) {
+	var result *C.struct_passwd
+	names := []string{}
+	// getpwent(3), posix compatible way to iterate /etc/passwd.
+	// Provided as os/user does not provide any iteration helpers
+	C.setpwent()
+	defer C.endpwent()
```

Comment:
`setpwent` and `getpwent` are `MT-Unsafe` (as per [setpwent(3)](https://man7.org/linux/man-pages/man3/setpwent.3.html)) on Linux and refer to thread-local storage on macOS, so we can't use them on either (we have zero control over which OS thread a goroutine ends up scheduled on).

Reply:
I'm not how to support macos in this case then as fgetpwent_r is a gnu specific extension that doesnt appear to be available on macos